From 775acf8e763ca77306c9a4e7997deb600a64ab97 Mon Sep 17 00:00:00 2001
From: Adriana Kobylak <anoo@us.ibm.com>
Date: Thu, 11 Oct 2018 12:12:06 -0500
Subject: [PATCH 3/3] msl_verify: Add support for multiple MSL values

There are scenarios where the BMC can support multiple PNOR MSL
values, such as v1.2 and v.1.4.9. Support multiple values
separated by a space.

Change-Id: Id65c43872db33538e2f02f6f60b4571a5231717b
Signed-off-by: Adriana Kobylak <anoo@us.ibm.com>
---
 msl_verify.cpp | 34 +++++++++++++++++++++-------------
 1 file changed, 21 insertions(+), 13 deletions(-)

diff --git a/msl_verify.cpp b/msl_verify.cpp
index e986c32..bc1585e 100644
--- a/msl_verify.cpp
+++ b/msl_verify.cpp
@@ -125,7 +125,7 @@ static void verify()
 
     constexpr auto purpose =
         "xyz.openbmc_project.Software.Version.VersionPurpose.Host";
-    auto min = std::string{PNOR_MSL};
+    auto minStr = std::string{PNOR_MSL};
 
     if (std::string(PNOR_MSL).empty())
     {
@@ -143,22 +143,30 @@ static void verify()
         return;
     }
 
-    Version_t min_t = {0, 0, 0};
-    parse(min, min_t);
-
     Version_t actual_t = {0, 0, 0};
     parse(actual, actual_t);
 
-    auto rc = compare(min_t, actual_t);
-    if (rc != 0)
+    std::istringstream minStream(minStr);
+    std::vector<std::string> mins(std::istream_iterator<std::string>{minStream},
+                                  std::istream_iterator<std::string>());
+
+    for (auto& min : mins)
     {
-        log<level::ERR>("PNOR Mininum Ship Level NOT met",
-                        entry("MIN_VERSION=%s", min.c_str()),
-                        entry("ACTUAL_VERSION=%s", actual.c_str()),
-                        entry("VERSION_PURPOSE=%s", purpose));
-        report<NotMet>(notMet::MIN_VERSION(min.c_str()),
-                       notMet::ACTUAL_VERSION(actual.c_str()),
-                       notMet::VERSION_PURPOSE(purpose));
+        Version_t min_t = {0, 0, 0};
+        parse(min, min_t);
+
+        auto rc = compare(min_t, actual_t);
+        if (rc != 0)
+        {
+            log<level::ERR>("PNOR Mininum Ship Level NOT met",
+                            entry("MIN_VERSION=%s", min.c_str()),
+                            entry("ACTUAL_VERSION=%s", actual.c_str()),
+                            entry("VERSION_PURPOSE=%s", purpose));
+            report<NotMet>(notMet::MIN_VERSION(min.c_str()),
+                           notMet::ACTUAL_VERSION(actual.c_str()),
+                           notMet::VERSION_PURPOSE(purpose));
+            break;
+        }
     }
 
     return;
-- 
1.8.3.1

